### PR TITLE
fix(job-server): Fix/partial fix for issue #876 - Performance (of) SqlDAO in SJS 0.7

### PR DESCRIPTION
**Pull request by Meehau**
* Separate the BINARIES table into BINARIES and BINARIES_CONTENTS, so that only the BINARIES_CONTENTS table contains the big "BINARY" (byte array) column. Both tables can be joined on the same BIN_ID (and sometimes are).
* Optimize the JobDAO.getLastUploadTimeAndType(...) method - made it abstract and implemented an optimized version in the JobFileDAO and JobSqlDAO.

**Pull Request checklist**
- [x] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :**
SqlDAO could use some performance improvements. See issue #876 

**New behavior :**
See commit comments above.

**BREAKING CHANGES**
None.

**Other information**:
None.

---
This change is [Reviewable](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/886)

### CODE REVIEW

File: job-server/src/test/scala/spark/jobserver/io/FlywayMigrationSpec.scala [hunk 1]
```diff
@@ -48,7 +48,8 @@ class FlywayMigrationSpec extends FunSpec with Matchers {
     val h2Pass = ""
     val h2Driver = "org.h2.Driver"
 
-    it("should migrate from JAR schema to more generic BINARY schema") {
+    it("should migrate from JAR schema to more generic BINARY schema " +
```
Comment (velvia) [inline job-server/src/test/scala/spark/jobserver/io/FlywayMigrationSpec.scala]: Would be nice to shorten this slightly

**Comment by thoHeinze** [comment]
@Meehau I think you need to rebase your change now, The PR #881 got merged into master.

**Comment by Meehau** [comment]
@thoHeinze Yes, unfortunately... I submitted this Pull Request 1.5 months ago, and in the meantime someone else's PR was accepted instead of this one (@msohn was also working on the performance issues). I don't know if this PR is even still viable. Ideally, this PR would get accepted and then @msohn could have worked on top of that, but now I've got to do some serious (precise) rebasing and merging, due to conflicts and the nontrivial SQL changes (like additional triggers for the PSQL case). Like, I haven't got any guarantees that I won't have to wait another 1.5 months for this carefully rebased and merged version to be reviewed/accepted again - I hope you guys understand that.

**Comment by noorul** [comment]
@Meehau Can you re-base please?

**Comment by thoHeinze** [comment]
@Meehau I'm sorry I'm not a contributor of this project, so I can't merge PR's. As discussed in the issue, @msohn  and I wanted as a next step to remove the duplicate storage of binaries. Your PR is a first step in this direction, so I would highly appreciate if we could get it in.

**Comment by velvia** [comment]
@Meehau apologies about the order of PR merges.  We are still interested in yours and as you can see I did review it and requested changes and a re-base....  if you have a chance to do so and update the PR we are happy to merge it.

**Comment by Meehau** [comment]
OK, since it seems that you guys find this PR to be rather important, I'll try to to rebase and merge it some time during this week.

**Comment by velvia** [comment]
Thanks Michal.

Evan Chan
velvia@gmail.com
----
“Never doubt that a small group of thoughtful, committed citizens can change the world.” —M. Mead

**Comment by msohn** [comment]
I rebased this pull request onto current master and reimplemented the migration since blob columns cannot be migrated using a script and created another pull request #932 for this change

**Comment by zeitos** [comment]
Hi @Meehau thanks for your contribution, do you think you'll be able to rebase?

**Comment by msohn** [comment]
I did this rebase already in my pull request #932 which is based on your pull request #886. Besides resolving conflicts I had to reimplement the migration as JDBC migration. This is necessary since blobs can't be accessed using SQL scripts

**Comment by velvia** [comment]
@Meehau @msohn thanks for the new PR #932.   Let's close this one and focus on the newer PR.

**Pull request closed without merging**